lib/net.js

```diff
@@ -631,10 +631,20 @@ Socket.prototype.destroySoon = function() {
   if (this.writable)
     this.end();
 
-  if (this.writableFinished)
+  if (this.writableFinished && this.readableEnded)
     this.destroy();
   else
-    this.once('finish', this.destroy);
+    this.once('finish', () => {
+      // If we destroy the socket before we have received an incoming EOF
+      // the remote might still be sending data that will trigger a RST
+      // from this host once it is received
+      // Closing a TCP socket requires two FIN/ACK exchanges, one in every
+      // direction - this means two events
+      if (this.readableEnded)
+        this.destroy();
+      else
+        this.once('end', this.destroy);
+    });
 };
```

Comment:
Where is destroySoon used? I think we were planning to eventually remove/deprecate this.

Reply:
HTTP(S) `res.end()` - almost exclusively

Reply:
What if you just replace `destroySoon` with `end`?

```js
  if (res._last) {
    if (typeof socket.destroySoon === 'function') {
      socket.destroySoon();
    } else {
      socket.end();
    }
  }
```

to

```js
  if (res._last) {
    socket.end();
  }
```